src/burnchains/bitcoin/indexer.rs

```diff
@@ -298,17 +396,24 @@ impl BitcoinIndexer {
             let msg_result = self.recv_message();
             match msg_result {
                 Ok(msg) => {
-                    // got a message, so handle it!
-                    let handled = self.handle_message(&msg, Some(message_handler));
+                    // got a message; go consume it
+                    let handled = self.handle_message(msg, Some(message_handler));
                     match handled {
                         Ok(do_continue) => {
                             keep_going = do_continue;
                             if !keep_going {
                                 debug!("Message handler indicates to stop");
                             }
                         }
-                        Err(btc_error::UnhandledMessage) => {
-                            debug!("Unhandled message {:?}", msg);
+                        Err(btc_error::UnhandledMessage(m)) => {
+                            match m.deref() {
+                                NetworkMessage::Alert(ref alert_msg) => {
+                                    keep_going = true;
```

Comment:
Is it possible for `keep_going` to be false at this point? it seem like `keep_going` can only be modified in the `Ok()` case.

Unrelatedly, if a ConnectionBroken occurs here, and `do_handshake` indicates that you need to re-establish a connection with the peer, is the session preserved by the bitcoin peer, or does it need to be re-initialized too?

Reply:
> Is it possible for keep_going to be false at this point? it seem like keep_going can only be modified in the Ok() case.

Yup, that's dead code.  We only handle `Alert` here because Bitcoin nodes have this annoying tendency to periodically tell us to upgrade (since we look like a pre-Segwit node to them).

> Unrelatedly, if a ConnectionBroken occurs here, and do_handshake indicates that you need to re-establish a connection with the peer, is the session preserved by the bitcoin peer, or does it need to be re-initialized too?

Nope.  `connect_handshake_backoff()` calls `self.connect()`, which calls `self.reconnect_peer()`, which re-initializes the socket (and the mutex guarding it).

Reply:
Right -- but do you need to call `begin_session` again?